dev-tools/reconfigurator-cli/src/main.rs

```diff
@@ -41,6 +46,14 @@ struct ReconfiguratorSim {
     /// blueprints created by the user
     blueprints: IndexMap<Uuid, Blueprint>,
 
+    /// external IPs allocated to services
+    ///
+    /// In the real system, external IPs have IDs, but those IDs only live in
+    /// CRDB - they're not part of the zone config sent from Reconfigurator to
+    /// sled-agent. This mimics the minimal bit of the CRDB `external_ip` table
+    /// we need.
+    external_ips: IndexMap<IpAddr, Uuid>,
```

Comment:
I'm trying to figure out the behavior of this -- like when and how it gets updated in `reconfigurator-cli`.  I see that we populate it when running `blueprint-plan` with a new Uuid if we haven't used this IP address before and we re-use the same id if we've seen this same IP address before.  Why is that correct?  Are we thinking of it like the parent blueprint has just been executed and we're starting from that state?  (I'm not seeing a problem here or anything -- just not quite getting it yet).

Will we need to have `omdb db reconfigurator-save` save this too?

Reply:
Good questions, and I admit I was kinda punting here (as I haven't actually used reconfigurator-cli in anger yet). For NICs we're in good shape: the network interface info included in `OmicronZoneType` includes the NIC ID, so we don't need to do anything special. But IPs are weird; `OmicronZoneType` includes the external IP for zones that need one, but does _not_ include the database ID of that external IP. I wonder if I'm putting a bandaid in the wrong place here, and instead we should send the external IP ID in the `OmicronZoneType` so we don't need to do anything at all here?

When I put this in, I was thinking of this as "some zones have external IPs; those have IDs that are effectively internal to CRDB, because they're not sent to the sleds as part of the zone config". To mimic that, I stashed them here.

> Will we need to have omdb db reconfigurator-save save this too?

Not sure. It may depend on how we use these in the planner, and how much execution-fakery reconfigurator-cli grows.

Reply:
> But IPs are weird; `OmicronZoneType` includes the external IP for zones that need one, but does _not_ include the database ID of that external IP. I wonder if I'm putting a bandaid in the wrong place here, and instead we should send the external IP ID in the `OmicronZoneType` so we don't need to do anything at all here?

I'm not sure.  But if we did that, would that be a better way to solve our other problem of knowing which records to be deleted when a zone gets expunged?  I'm not actually sure it's any better or even if it would work.

I think this is probably fine for now and we may rethink it as we iterate on `reconfigurator-cli` -- its model of "state outside the blueprint" is getting a little creaky in other ways, too.

Reply:
> I'm not sure. But if we did that, would that be a better way to solve our other problem of knowing which records to be deleted when a zone gets expunged? I'm not actually sure it's any better or even if it would work.

No, we'd still have to fetch all the external IPs for this to work. But we wouldn't need this extra state in `reconfigurator-cli` just to describe the in-service zones.

I'll go with "fine for now, iterate later", though. :-)